### PR TITLE
Introduce Waf package

### PR DESCRIPTION
**Pull request by miguelxpn**
#### Changes proposed in this Pull Request:

* Introducing the WAF package with the rule engine.

#### Jetpack product discussion

pbuNQi-1T7-p2

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Run the package unit tests

### CODE REVIEW

**Review by bhoop — CHANGES_REQUESTED** [review]
Thanks for getting this started! Do you think it would make sense to add some functionality here to actually utilize the rule engine - i.e. have a simple rule for a specific request that we can use to confirm that this is all working? Maybe behind a wp_options flag that defaults to "off" but we could turn on manually in the database or with a wp-cli command?

**Comment by miguelxpn** [comment]
> Thanks for getting this started! Do you think it would make sense to add some functionality here to actually utilize the rule engine - i.e. have a simple rule for a specific request that we can use to confirm that this is all working? Maybe behind a wp_options flag that defaults to "off" but we could turn on manually in the database or with a wp-cli command?

I think it makes more sense to have the code that handles requests in the plugin. That way we don't couple this package with the request handling and whoever consumes it can use it in their own specific way to handle the request.

> Do we need to do something with these two lines?

According to the monorepo [documentation](https://github.com/Automattic/jetpack/blob/master/docs/monorepo.md#creating-a-new-composer-package) the Crew team will add the composer.json file, so perhaps they will adjust it for us? I'll ask them and find out!

**Comment by jeherve** [comment]
Hi team!

As you work on this functionality, I would recommend that you aim to merge this first PR as early as possible. It doesn't have to be all ready before you merge it. If we can merge a first, smaller PR, early on, and then iterate, feature by feature, in multiple small PRs, it will be easier to review for everyone.

Thank you!

**Review by bhoop — APPROVED** [review]
Per @jeherve's point, let's get this merged and start iterating on it.

**Review by bhoop — APPROVED** [review]
👍

File: projects/packages/waf/composer.json [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		"classmap": [
+			"src/"
+		]
+	},
+	"scripts": {
+		"phpunit": [
+			"./vendor/phpunit/phpunit/phpunit --colors=always"
+		],
+		"test-coverage": [
+			"@composer install",
```
Comment (anomiex) [inline projects/packages/waf/composer.json:21]: Remove this to match #22390
```suggestion
```

File: projects/packages/waf/package.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"license": "GPL-2.0-or-later",
+	"author": "Automattic",
+	"scripts": {
+		"build": "echo 'Not implemented.",
+		"build-js": "echo 'Not implemented.",
+		"build-production": "echo 'Not implemented.",
+		"build-production-js": "echo 'Not implemented.",
+		"clean": "true",
+		"distclean": "rm -rf node_modules && pnpm run clean",
+		"install-if-deps-outdated": "pnpm install --no-prod --frozen-lockfile"
```
Comment (anomiex) [inline projects/packages/waf/package.json:23]: Remove this to match #22390
```suggestion
		"clean": "true"
```
Or just remove the whole file if there isn't going to be any JS in the package.

**Review by anomiex — COMMENTED** [review]
Gave it a once-over from a monorepo structural perspective. Didn't review the actual code.

File: projects/packages/waf/src/class-multi-string-matcher.php [hunk 3]
```diff
@@ -0,0 +1,232 @@
+<?php // @codingStandardsIgnoreStart
```
[the remark this answers is not in the document]
Comment (miguelxpn) [reply projects/packages/waf/src/class-multi-string-matcher.php:1]: This version has a fix that @bhoop wrote. We'll try to get the fix upstream, once it's there we'll remove the file from this package and import it from packagist.

[the remark this answers is not in the document]
Comment (anomiex) [reply projects/packages/waf/src/class-multi-string-matcher.php:1]: If you get a patch into their Gerrit system and it doesn't get attention, let me know and I can try to ping some people I know there.

[the remark this answers is not in the document]
Comment (miguelxpn) [reply projects/packages/waf/src/class-multi-string-matcher.php]: @anomiex I already got the patch on their Gerrit system and updated the PR to pull the dependency from composer. I think we can merge this as is now and then update it later once the patch lands if that's ok with you.

**Review by jeherve — APPROVED** [review]
Let's merge this! 🚢

File: projects/packages/waf/src/class-multi-string-matcher.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (anomiex) [reply projects/packages/waf/src/class-multi-string-matcher.php]: Fine with me

**Pull request merged**